File: SymmetricTree_DFS.java
/* Time Complexity: O(n)
 * Space Complexity: O(h), h is the height of the tree, recursive stack space
 * Did this code successfully run on Leetcode : yes
 * Any problem you faced while coding this : No
 */

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        return dfs(root.left, root.right);
    }
    private boolean dfs(TreeNode left, TreeNode right){
        //base
        if(left == null && right == null) return true;
        if(left == null || right == null || left.val != right.val) return false;
        //logic
        return dfs(left.left, right.right) && dfs(left.right, right.left);
    }
}

File: symmetric_tree_dfs.py
"""
Time Complexity : O(n) where n is the total number of nodes in the tree
Space Complexity : O(n). In worst case scenario all the nodes can be push to recursive stack

Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No
"""
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:

    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        if root == None:
            return True
        return self.dfs(root.left, root.right)

    def dfs(self, left, right):
        # base case
        if left == None and right == None:
            return True
        if left == None and right != None:
            return False
        if left != None and right == None:
            return False
        if left.val != right.val:
            return False

        # logic
        return self.dfs(left.left, right.right) and self.dfs(left.right, right.left)
